File: internal/cli/start_test.go
package cli

import (
	"bytes"
	"fmt"
	"os"
	"strings"
	"testing"

	"harness/internal/orchestrator"
	"harness/internal/runner/runnertest"
)

func TestStartArgumentsParseInAnyOrder(t *testing.T) {
	for _, args := range [][]string{
		{"hx", "--max", "2", "--foreground"},
		{"--max", "2", "hx", "--foreground"},
		{"--foreground", "--max", "2", "hx"},
		{"--foreground", "hx", "--max=2"},
	} {
		got, err := parseStart(args, &strings.Builder{})
		if err != nil || got.epic != "hx" || got.max != 2 || !got.foreground {
			t.Errorf("parseStart(%q) = %+v, %v", args, got, err)
		}
	}
	if got, err := parseStart([]string{"--ticket", "hx-1"}, &strings.Builder{}); err != nil || got.ticket != "hx-1" || got.max != 3 {
		t.Errorf("--ticket: %+v, %v", got, err)
	}
	for _, bad := range [][]string{{}, {"hx", "--ticket", "hx-1"}, {"hx", "extra"}, {"hx", "--max", "0"}} {
		if _, err := parseStart(bad, &strings.Builder{}); err == nil {
			t.Errorf("parseStart(%q) accepted", bad)
		}
	}
	// The detached child must see --foreground whatever the user's argument order.
	if got := fmt.Sprint(foregroundArgs([]string{"start", "--max", "2", "hx"})); got != "[start --foreground --max 2 hx]" {
		t.Errorf("foregroundArgs = %s", got)
	}
}

func TestSecondStartInTheSameRepoRefuses(t *testing.T) {
	repo := preparedRepo(t)
	release, err := orchestrator.AcquireLock(repo) // a live Orchestrator: this process
	if err != nil {
		t.Fatal(err)
	}
	defer release()
	var out bytes.Buffer
	code := Run([]string{"start", "hx", "--foreground"}, &out, repo, (&runnertest.Fake{Handle: okTools}).Run, herdrEnv)
	if code == 0 || !strings.Contains(out.String(), "already running") {
		t.Errorf("second start: exit %d, output %q", code, out.String())
	}
}

func TestControlCommandsReachOnlyARunningOrchestrator(t *testing.T) {
	repo := t.TempDir()
	var out bytes.Buffer
	if code := Run([]string{"stop"}, &out, repo, nil, herdrEnv); code == 0 || !strings.Contains(out.String(), "no Orchestrator is running") {
		t.Errorf("stop with nothing running: exit %d, %q", code, out.String())
	}

	release, _ := orchestrator.AcquireLock(repo)
	defer release()
	for args, file := range map[string]string{"stop": "stop", "retry hx-1": "retry-hx-1", "park hx-1": "park-hx-1", "address hx-1": "address-hx-1"} {
		if code := Run(strings.Fields(args), &out, repo, nil, herdrEnv); code != 0 {
			t.Errorf("%s: exit %d", args, code)
		}
		if _, err := os.Stat(orchestrator.ControlFile(repo, file)); err != nil {
			t.Errorf("%s left no control file %s", args, file)
		}
	}
	if code := Run([]string{"retry"}, &out, repo, nil, herdrEnv); code == 0 {
		t.Errorf("retry without a Ticket accepted")
	}
}

File: src/cli/start_test.rs
use super::init_test::{herdr_env, ok_tools, prepared_repo, run_with};
use super::{parse_start, StartArgs};
use crate::orchestrator::stage::control_file;
use crate::orchestrator::state::acquire_lock;
use crate::tempdir::TempDir;
use crate::tools::fake::Fake;

fn parse(args: &[&str]) -> Result<StartArgs, String> {
    let args: Vec<String> = args.iter().map(|a| a.to_string()).collect();
    parse_start(&args)
}

#[test]
fn start_arguments_parse_in_any_order() {
    for args in [
        &["hx", "--max", "2"][..],
        &["--max", "2", "hx"],
        &["hx", "--max=2"],
        &["-max", "2", "hx"],
    ] {
        let got = parse(args);
        assert_eq!(
            got,
            Ok(StartArgs {
                epic: "hx".to_string(),
                ticket: String::new(),
                max: 2
            }),
            "parse_start({args:?})"
        );
    }
    let got = parse(&["--ticket", "hx-1"]).unwrap();
    assert!(got.ticket == "hx-1" && got.max == 3, "--ticket: {got:?}");
    // A run stays in the pane it was started from: --detach is gone (ADR 0003).
    for bad in [
        &[][..],
        &["hx", "--ticket", "hx-1"],
        &["hx", "extra"],
        &["hx", "--max", "0"],
        &["hx", "--detach"],
    ] {
        assert!(parse(bad).is_err(), "parse_start({bad:?}) accepted");
    }
    // A bad flag is named before the usage, as Go's flag package did.
    let (code, out) = run_with(
        &["start", "hx", "--max", "lots"],
        prepared_repo().path(),
        ok_tools(),
        &herdr_env,
    );
    assert!(
        code == 2 && out.starts_with("invalid value \"lots\" for flag -max\nusage:"),
        "bad --max: exit {code}, output {out:?}"
    );
}

#[test]
fn second_start_in_the_same_repo_refuses() {
    let repo = prepared_repo();
    let _lock = acquire_lock(repo.path()).unwrap(); // a live Orchestrator: this process
    let (code, out) = run_with(&["start", "hx"], repo.path(), ok_tools(), &herdr_env);
    assert!(
        code != 0 && out.contains("already running"),
        "second start: exit {code}, output {out:?}"
    );
}

#[test]
fn control_commands_reach_only_a_running_orchestrator() {
    let repo = TempDir::new();
    let (code, out) = run_with(&["stop"], repo.path(), Fake::quiet(), &herdr_env);
    assert!(
        code != 0 && out.contains("no Orchestrator is running"),
        "stop with nothing running: exit {code}, {out:?}"
    );

    let _lock = acquire_lock(repo.path()).unwrap();
    for (args, file) in [
        (&["stop"][..], "stop"),
        (&["retry", "hx-1"], "retry-hx-1"),
        (&["park", "hx-1"], "park-hx-1"),
        (&["address", "hx-1"], "address-hx-1"),
    ] {
        let (code, out) = run_with(args, repo.path(), Fake::quiet(), &herdr_env);
        assert_eq!(code, 0, "{args:?}: exit {code}, output {out:?}");
        assert!(
            control_file(repo.path(), file).exists(),
            "{args:?} left no control file {file}"
        );
    }
    let (code, _) = run_with(&["retry"], repo.path(), Fake::quiet(), &herdr_env);
    assert_ne!(code, 0, "retry without a Ticket accepted");
}
